sotodlib/site_pipeline/finalize_focal_plane.py

```diff
@@ -214,45 +409,224 @@ def _gamma_min(scale, shift, gamma):
     return res.x
 
 
-def to_horiz(points, encoders):
-    """
-    Go from xieta coordinates to horizon coordinates.
+def _load_template(template_path, ufm, pointing_cfg):
+    template_rset = read_dataset(template_path, ufm)
+    det_ids = template_rset["dets:det_id"]
+    template = np.column_stack(
+        (
+            np.array(template_rset["xi"]),
+            np.array(template_rset["eta"]),
+            np.array(template_rset["gamma"]),
+        )
+    )
+    template_optical = template_rset["is_optical"]
 
-    Arguments:
+    return Template(
+        np.array(det_ids), template.T, np.array(template_optical), pointing_cfg
+    )
 
-        points: (3, ndet) array of points.
-                Rows should be (xi, eta, gamma).
 
-        encoders: The encoder positions for the LOS.
-                  Should be (az, el, bs).
+def _load_ctx(config):
+    ctx = Context(config["context"]["path"])
+    tod_pointing_name = config["context"].get("tod_pointing", "tod_pointing")
+    map_pointing_name = config["context"].get("map_pointing", "map_pointing")
+    pol_name = config["context"].get("polarization", "polarization")
+    dm_name = config["context"].get("detmap", "detmap")
+    query = []
+    if "query" in config["context"]:
+        query = (ctx.obsdb.query(config["context"]["query"])["obs_id"],)
+    obs_ids = np.append(config["context"].get("obs_ids", []), query)
+    obs_ids = np.unique(obs_ids)
+    if len(obs_ids) == 0:
+        raise ValueError("No observations provided in configuration")
+    _config = config.copy()
+    if "query" in _config["context"]:
+        del _config["context"]["query"]
+    amans = []
+    dets = config["context"].get("dets", {})
+    for obs_id in obs_ids:
+        aman = ctx.get_meta(obs_id, dets=dets)
+        if "wafer" not in aman.det_info and dm_name in aman:
+            dm_aman = aman[dm_name].copy()
+            aman.det_info.wrap("wafer", dm_aman)
+            if "det_id" not in aman.det_info:
+                aman.det_info.wrap(
+                    "det_id", aman.det_info.wafer.det_id, [(0, aman.dets)]
+                )
+        if "det_id" in aman.det_info:
+            aman.restrict("dets", aman.dets.vals[aman.det_info.det_id != ""])
+            aman.restrict("dets", aman.dets.vals[aman.det_info.det_id != "NO_MATCH"])
+        elif "det_info" not in aman or "det_id" not in aman.det_info:
```

Comment:
If you want to trap this informatively, it needs to come first.  Currently it's equivalent to "else" (code will already be crashed if det_info not in aman).